demos/keyboard_nav/line_cursor.js

```diff
@@ -0,0 +1,177 @@
+ * @return {Blockly.ASTNode} The next node, or null if the current node is
+ *     not set or there is no next value.
+ * @override
+ */
+Blockly.LineCursor.prototype.next = function() {
+  var curNode = this.getCurNode();
+  if (!curNode) {
+    return null;
+  }
+  var newNode = this.getNextNode_(curNode, this.validNode_);
+
+  // Skip the input or next value if there is a connected block.
+  if (newNode && (newNode.getType() == Blockly.ASTNode.types.INPUT ||
+      newNode.getType() == Blockly.ASTNode.types.NEXT) &&
+      newNode.getLocation().targetBlock()) {
+    newNode = this.getNextNode_(newNode, this.validNode_);
+  }
+  if (newNode) {
+    this.setCurNode(newNode);
+  }
+  return newNode;
+};
+
+/**
+ * For a basic cursor we only have the ability to go next and previous, so
+ * in will also allow the user to get to the next node in the pre order traversal.
+ * @return {Blockly.ASTNode} The next node, or null if the current node is
+ *     not set or there is no next value.
+ * @override
+ */
+Blockly.LineCursor.prototype.in = function() {
+  var curNode = this.getCurNode();
+  if (!curNode) {
+    return null;
+  }
+  var newNode = this.getNextNode_(curNode, this.validInNode_);
+
+  if (newNode) {
+    this.setCurNode(newNode);
+  }
+  return newNode;
+};
+
+/**
+ * Find the previous node in the pre order traversal.
+ * @return {Blockly.ASTNode} The previous node, or null if the current node
+ *     is not set or there is no previous value.
+ * @override
+ */
+Blockly.LineCursor.prototype.prev = function() {
+  var curNode = this.getCurNode();
+  if (!curNode) {
+    return null;
+  }
+  var newNode = this.getPreviousNode_(curNode, this.validNode_);
+  
+  if (newNode && (newNode.getType() == Blockly.ASTNode.types.INPUT ||
+    newNode.getType() == Blockly.ASTNode.types.NEXT) &&
+    newNode.getLocation().targetBlock()) {
+    newNode = this.getPreviousNode_(newNode, this.validNode_);
+  }
+
+  if (newNode) {
+    this.setCurNode(newNode);
+  }
+  return newNode;
+};
+
+/**
+ * For a basic cursor we only have the ability to go next and previou, so
+ * out will allow the user to get to the previous node in the pre order traversal.
+ * @return {Blockly.ASTNode} The previous node, or null if the current node is
+ *     not set or there is no previous value.
+ * @override
+ */
+Blockly.LineCursor.prototype.out = function() {
+  var curNode = this.getCurNode();
+  if (!curNode) {
+    return null;
+  }
+  var newNode = this.getPreviousNode_(curNode, this.validInNode_);
+
+  if (newNode) {
+    this.setCurNode(newNode);
+  }
+  return newNode;
+
+};
+
+/**
+ * Decides what nodes to traverse and which ones to skip. Currently, it
+ * skips output, stack and workspace nodes.
+ * @param {Blockly.ASTNode} node The AST node to check whether it is valid.
+ * @return {boolean} True if the node should be visited, false otherwise.
+ * @private
+ */
+Blockly.LineCursor.prototype.validNode_ = function(node) {
```

Comment:
Can we come up with a better name for validNode_ and validInNode_, its hard to tell what the difference is without reading comments.

Reply:
We can but I am struggling haha.  validLineNode and validInLineNode?

Reply:
Sounds good.